### PR TITLE
tools/funclatency: support nested or recursive functions

### PR DESCRIPTION
**Pull request by chenyuezhou**
# Implementation

add `-l level` option to support nested or recursive functions.

without `-l` option, funclatency tool worked as before.

```
typedef struct {
    u64 ip;
    u64 start_ts;
} func_cache_t;

/* LIFO */
typedef struct {
    u32          head;
    func_cache_t cache[STACK_DEPTH];
} func_stack_t;

BPF_HASH(func_stack, u32, func_stack_t);
```

cache stack(LIFO) per pid.

### CODE REVIEW

**Comment by yonghong-song** [comment]
[buildbot, test this please]

File: tools/funclatency.py [hunk 1]
```diff
@@ -62,6 +57,8 @@
     help="show a separate histogram per function")
 parser.add_argument("-r", "--regexp", action="store_true",
     help="use regular expressions. Default is \"*\" wildcards only.")
+parser.add_argument("-l", "--level", type=int,
+    help="set the level of nested or recursive functions")
```
Comment (yonghong-song) [inline tools/funclatency.py:61]: Looks great! Please change example file and man page as well since we added a new option here.

**Comment by brendangregg** [comment]
Do you have any idea of the overhead the -l option causes? Thanks for working on this as I think we do need the option, but if there's added overhead we should mention it in the man page, ideally with a rough figure as a guide

File: tools/funclatency.py [hunk 2]
```diff
@@ -12,11 +12,6 @@
 # The pattern is a string with optional '*' wildcards, similar to file
 # globbing. If you'd prefer to use regular expressions, use the -r option.
 #
-# Currently nested or recursive functions are not supported properly, and
-# timestamps will be overwritten, creating dubious output. Try to match single
-# functions, or groups of functions that run at the same stack layer, and
-# don't ultimately call each other.
-#
 # Copyright (c) 2015 Brendan Gregg.
```
Comment (yonghong-song) [inline tools/funclatency.py:15]: This is rather a big change. You can add a "Copyright" note for your contribution here if you want. Thanks!

**Comment by chenyuezhou** [comment]
I tested it by calling a function with a specified nesting level for million times.

Total1(without funclatency): Average time.

Total2(with funclatency): Average time.

Increase: Total2 - Total1.

(when nesting depth is more than 31, BPF stack limit of 512 bytes is exceeded.)

|  Nesting depth   |  Total1(ns) | Total2(ns) | Increase(ns) |
|  ----  | ----  | ---- | ---- |
| 2  | 31.1004 | 3996.8870 | 3965.7866 |
| 4  | 63.8534 | 8015.8878 | 7952.0344 |
| 8  | 121.3456 | 16402.8414 | 16281.4958 |
| 16 | 268.4610 | 32237.9625 | 31969.5015 |

File: tools/funclatency.py [hunk 3]
```diff
@@ -178,14 +171,140 @@ def bail(error):
...
+            """
+    hist_key_t key;
+    key.key.ip  = ip;
+    key.key.pid = %s;
+    key.slot    = bpf_log2l(delta);
+    dist.increment(key);
+
+    if (stack->head == 0) {
+        /* empty */
+
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply tools/funclatency.py]: fixed.

File: tools/funclatency.py [hunk 4]
```diff
@@ -178,14 +171,140 @@ def bail(error):
...
+    if (!stack) {
+        /* miss start */
+
+        return 0;
+    }
+
+    func_cache_t cache = {};
+    if (stack_pop(stack, &cache)) {
+        func_stack.delete(&pid);
+
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply tools/funclatency.py:275]: fixed.

File: tools/funclatency.py [hunk 5]
```diff
@@ -178,14 +171,140 @@ def bail(error):
...
+    }
+            """)
+
+        bpf_text = bpf_text.replace('CALCULATE',
+            """
+    u64 ip, start_ts;
+    func_stack_t *stack = func_stack.lookup(&pid);
+    if (!stack) {
+        /* miss start */
+
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply tools/funclatency.py]: fixed.

File: tools/funclatency.py [hunk 6]
```diff
@@ -178,14 +171,140 @@ def bail(error):
...
+    u32 index = stack->head;
+
+    if (index > STACK_DEPTH - 1) {
+        /* bound check */
+
+        return -1;
+    }
+
+    stack->head++;
+
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply tools/funclatency.py]: fixed.

File: tools/funclatency.py [hunk 7]
```diff
@@ -178,14 +171,140 @@ def bail(error):
...
+    return 0;
+}
+
+static inline int stack_push(func_stack_t *stack, func_cache_t *cache) {
+    u32 index = stack->head;
+
+    if (index > STACK_DEPTH - 1) {
+        /* bound check */
+
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply tools/funclatency.py]: fixed

File: tools/funclatency.py [hunk 8]
```diff
@@ -178,14 +171,140 @@ def bail(error):
...
-    bpf_text = bpf_text.replace('STORAGE', 'BPF_HASH(ipaddr, u32);\n' +
-        'BPF_HISTOGRAM(dist, hist_key_t);')
-    # stash the IP on entry, as on return it's kretprobe_trampoline:
-    bpf_text = bpf_text.replace('ENTRYSTORE',
-        'u64 ip = PT_REGS_IP(ctx); ipaddr.update(&pid, &ip);')
     pid = '-1' if not library else 'tgid'
-    bpf_text = bpf_text.replace('STORE',
-        """
+
+    if args.level > 1:
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply tools/funclatency.py]: fixed

File: tools/funclatency.py [hunk 9]
```diff
@@ -178,14 +171,140 @@ def bail(error):
...
+
+        bpf_text = bpf_text.replace('FUNCTION',
+            """
+static inline int stack_pop(func_stack_t *stack, func_cache_t *cache) {
+    if (stack->head <= 0) {
+        return -1;
+    }
+
+    u32 index = --stack->head;
+
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply tools/funclatency.py]: fixed

File: tools/funclatency.py [hunk 10]
```diff
@@ -12,11 +12,6 @@
 # The pattern is a string with optional '*' wildcards, similar to file
 # globbing. If you'd prefer to use regular expressions, use the -r option.
 #
-# Currently nested or recursive functions are not supported properly, and
-# timestamps will be overwritten, creating dubious output. Try to match single
-# functions, or groups of functions that run at the same stack layer, and
-# don't ultimately call each other.
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply tools/funclatency.py:18]: added.

File: tools/funclatency.py (hunk 1 above)
Comment (chenyuezhou) [reply tools/funclatency.py:66]: Done.

**Comment by brendangregg** [comment]
> cpu bogomips: 3993.60
> 
> I tested it by calling a function with a specified nesting level for million times.
> 
> (when nesting depth is more than 31, BPF stack limit of 512 bytes is exceeded.)
> Nesting depth 	Avg1(ns) 	Avg2(ns) 	Overhead(ns)
> 2 	31.1004 	3996.8870 	3965.7866
> 4 	63.8534 	8015.8878 	7952.0344
> 8 	121.3456 	16402.8414 	16281.4958
> 16 	268.4610 	32237.9625 	31969.5015

Thanks, but the real question I have is what's the extra overhead between using the -l option and not, for regular (non-nested) functions, so the man page when describing the -l option can note extra overhead if that happens.

**Comment by chenyuezhou** [comment]
run `dd if=/dev/zero of=/dev/null bs=1 count=10485760`

```
10485760 bytes (10 MB, 10 MiB) copied, 10.0763 s, 1.0 MB/s
```

run `dd if=/dev/zero of=/dev/null bs=1 count=10485760` and `funclatency vfs_read`

```
10485760 bytes (10 MB, 10 MiB) copied, 18.2491 s, 575 kB/s
```

run `dd if=/dev/zero of=/dev/null bs=1 count=10485760` and `funclatency vfs_read -l 2`

```
10485760 bytes (10 MB, 10 MiB) copied, 18.2299 s, 575 kB/s
```

In this test the overhead between using the `-l` option and not is negligible.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
From implementation perspective, the new code is generated only under level > 1, so we should not see a degradation of performance for level <= 1.

**Pull request merged**